blog/views.py

```diff
@@ -3,115 +3,136 @@
+
+
+class EditPost(UpdateView, Protected):
+    """View for editing post."""
+
+    model = Post
+    fields = ['title', 'text']
+    template_name = 'blog/post_edit.html'
+
+    def get_success_url(self):
+        """Return to detail view of edited post."""
+        return reverse_lazy('post_detail', kwargs={'pk': self.object.pk})
+
+
+class PostDraftList(ListView, Protected):
+    """View for list with unpublished posts."""
+
+    context_object_name = 'posts'
+    template_name = 'blog/post_draft_list.html'
+    queryset = Post.objects.filter(published_date__isnull=True).order_by('created_date')
+
+
+class PublishPost(TemplateView, Protected):
+    """View for publishing post."""
+
+    def get(self, request, *args, **kwargs):
+        """Publish post."""
+        post = get_object_or_404(Post, pk=kwargs['pk'])
+        post.publish()
+        return redirect('post_detail', pk=post.pk)
+
+
+class RemovePost(DeleteView, Protected):
+    """View for deleting post."""
+
+    model = Post
+    success_url = reverse_lazy('post_list')
+    template_name = 'blog/post_edit.html'
+
+
+class AddComment(CreateView):
+    """View for adding comment to post."""
+
+    model = Comment
+    fields = ['author', 'text']
+    template_name = 'blog/add_comment.html'
+
+    def post(self, request, *args, **kwargs):
+        """Save new comment to DB."""
+        post = get_object_or_404(Post, pk=kwargs['pk'])
         form = CommentForm(request.POST)
         if form.is_valid():
             comment = form.save(commit=False)
             comment.post = post
             comment.save()
             return redirect('post_detail', pk=post.pk)
-    else:
-        form = CommentForm()
-    return render(request, 'blog/add_comment.html', {'form': form})
-
-
-@login_required
-def comment_approve(request, pk):
-    """Approve (publish) comment."""
-    comment = get_object_or_404(Comment, pk=pk)
-    comment.approve()
-    return redirect('post_detail', pk=comment.post.pk)
-
-
-@login_required
-def comment_remove(request, pk):
-    """Delete comment."""
-    comment = get_object_or_404(Comment, pk=pk)
-    post_pk = comment.post.pk
-    comment.delete()
-    return redirect('post_detail', pk=post_pk)
+
+
+class ApproveComment(TemplateView, Protected):
+    """View for approving comment, if authorized."""
+
+    def get(self, request, *args, **kwargs):
+        """Approve comment."""
+        comment = get_object_or_404(Comment, pk=kwargs['pk'])
+        comment.approve()
+        return redirect('post_detail', pk=comment.post.pk)
+
+
+class RemoveComment(DeleteView, Protected):
+    """View for deleting comment, if authorized."""
+
+    model = Comment
+    template_name = 'blog/post_detail.html'
+
+    def get(self, request, *args, **kwargs):
+        """Delete comment and return to base post."""
+        comment = get_object_or_404(Comment, pk=kwargs['pk'])
+        post_pk = comment.post.pk
+        comment.delete()
+        return redirect('post_detail', pk=post_pk)
```

Comment:
How about get_success_url?

Reply:
Tried, but didn't figure out, how to get post.pk. 
get_success_url evaluates after the action is done, so I can't get it from self.object.

Reply:
I'd save it in form_valid and then reused that to create success url

Reply:
Actually, I was wrong - we can get self.object.
https://code.djangoproject.com/ticket/19044

But right now it fails on template rendering (cuz post object is missing). I can write separate template for deleting or redefine get method (weird solution, I guess)...

Reply:
Alright. Let's leave it as it. It would be cleaner in case of doing normal CRUD for some JS client.